truffle/src/main/java/org/jruby/truffle/nodes/ext/BigDecimalNodes.java

```diff
@@ -482,194 +691,113 @@ public NegNode(RubyContext context, SourceSection sourceSection) {
-                default:
-                    throw new RuntimeException(); // never reached
-
             }
+            CompilerAsserts.neverPartOfCompilation();
+            throw new UnsupportedOperationException();
         }
 
     }
 
-    @CoreMethod(names = "sub", required = 2)
-    public abstract static class SubNode extends BigDecimalCoreMethodNode {
-
-        public SubNode(RubyContext context, SourceSection sourceSection) {
-            super(context, sourceSection);
-        }
-
-        @TruffleBoundary
-        private Object subBigDecimal(RubyBasicObject a, BigDecimal b, int precision) {
-            return createRubyBigDecimal(getBigDecimalValue(a).subtract(b, new MathContext(precision)));
-        }
-
-        @Specialization(guards = "isNormal(a)")
-        public Object sub(RubyBasicObject a, long b, int precision) {
-            return subBigDecimal(a, getBigDecimalValue(b), precision);
-        }
-
-        @Specialization(guards = "isNormal(a)")
-        public Object sub(RubyBasicObject a, double b, int precision) {
-            return subBigDecimal(a, getBigDecimalValue(b), precision);
-        }
-
-        @Specialization(guards = {"isNormal(a)", "isRubyBignum(b)"})
-        public Object subBignum(RubyBasicObject a, RubyBasicObject b, int precision) {
-            return subBigDecimal(a, getBignumBigDecimalValue(b), precision);
-        }
-
-        @Specialization(guards = {
-                "isNormal(a)",
-                "isRubyBigDecimal(b)",
-                "isNormal(b)"})
-        public Object sub(RubyBasicObject a, RubyBasicObject b, int precision) {
-            return subBigDecimal(a, getBigDecimalValue(b), precision);
-        }
-    }
-
-    @CoreMethod(names = "*", required = 1)
-    public abstract static class MultOpNode extends BigDecimalCoreMethodNode {
+    public abstract static class AbstractMultNode extends OpNode {
 
         private final ConditionProfile zeroNormal = ConditionProfile.createBinaryProfile();
 
-        public MultOpNode(RubyContext context, SourceSection sourceSection) {
+        public AbstractMultNode(RubyContext context, SourceSection sourceSection) {
             super(context, sourceSection);
         }
 
         @TruffleBoundary
-        private Object multBigDecimal(RubyBasicObject a, BigDecimal b) {
-            if (zeroNormal.profile(isNormalZero(a) && b.signum() == -1))
-                return createRubyBigDecimal(Type.NEGATIVE_ZERO);
-            return createRubyBigDecimal(getBigDecimalValue(a).multiply(b));
-        }
+        private Object multBigDecimal(RubyBasicObject a, RubyBasicObject b, MathContext mathContext) {
+            final BigDecimal bBigDecimal = getBigDecimalValue(b);
 
-        @Specialization(guards = "isNormal(a)")
-        public Object mult(RubyBasicObject a, long b) {
-            return multBigDecimal(a, getBigDecimalValue(b));
-        }
+            if (zeroNormal.profile(isNormalZero(a) && bBigDecimal.signum() == -1))
```

Comment:
Good catch only the `BigDecimal#multiply` should be behind the boundary. Extracted.